// scraping/scrapers/bigw.js
import { filterParams, setParams } from 'scraper.js';

const ONE_DAY = 86400000;

export default {
  start: {
    url: 'https://www.bigw.com.au/',
    method: 'category'
  },
  timeBetween: ONE_DAY,
  name: 'BigW',
  category({$, queue}) {
    $('a[href*="/c/"]')
      .map((i, x) => $(x).prop('href')).get()
      .forEach(href => queue(href, 'category'));
    $('a[href*="/p/"]')
      .map((i, x) => $(x).prop('href')).get()
      .forEach(href => queue(href, 'product', {priority: 'high', attempts: 3, backoff: 3600000}));
  },
  product({$, data}) {
    data({
      name: $('h1[itemprop="name"]').text().trim(),
      price: $('[itemprop="price"]:contains($):first').text().trim(),
      image: $('.MagicZoomPlus').prop('href'),
      description: $('[itemprop="description"]').text().trim(),
    });
  },
  filterQueueItem(queueItem) {
    if (queueItem.method === 'category') {
      return setParams(filterParams(queueItem.url, ['page']), {
        pageSize: 144,
      });
    }
    return queueItem.url;
  }
};
