// scraping/scrapers/jb-hifi.js
import { filterParams, setParams } from 'scraper.js';
const ONE_DAY = 86400000;

export default {
  name: 'JB Hi-Fi',
  timeBetween: ONE_DAY,
  start: {
    url: 'https://www.jbhifi.com.au/',
    method: 'home'
  },
  home({$, queue}) {
    $('ul.products a')
      .map((i, x) => $(x).prop('href')).get()
      .forEach(href => queue(href, 'category'));
  },
  category({$, queue}) {
    $('[data-productid] a.link')
      .map((i, x) => $(x).prop('href')).get()
      .forEach(href => queue(href, 'product', {priority: 'high', attempts: 3, backoff: 3600000}));
    $('.pagNum a')
      .map((i, x) => $(x).prop('href')).get()
      .forEach(href => queue(href, 'category'));
  },
  product({$, data}) {
    data({
      name: $('.primary h1').text().trim(),
      price: $('meta[property=price]').attr('content'),
      image: window.location.protocol + '//' + window.location.hostname + $('.gallery .image img').not('.special').attr('data-src-retina'),
      description: $('[itemprop=description] > *').not('.wishlistBtnContainer, #BVRRSummaryContainer').text(),
      brand: $('.overview .brand img').attr('alt'),
    });
  },
  filterQueueItem(queueItem) {
    if (queueItem.method === 'category') {
      return setParams(filterParams(queueItem.url, ['pg']));
    }
    return queueItem.url;
  }
}
